### DISCUSSION TITLE
Creating data-driven labels

### DISCUSSION THREAD

**Discussion by zmatez** [maplibre/maplibre-gl-js#4938] [General]
Since feature-state doesn't support layout, how can I dynamically add properties to display on the map?
Like:
```ts
            map.addLayer({
                id: 'my-label-layer',
                type: 'symbol',
                source: 'my-data-source',
                "source-layer": "terrain_wojewodztwa_centroid",
                layout: {
                    'text-field': [
                        'format',
                        ['get', 'jpt_nazwa_'], { 'font-scale': 1.0 },
                        '\ntest:', {},
                        ['get', 'external'], { 'font-scale': 0.8, 'text-color': '#666' }
                    ],
                    "text-font": ["Noto Sans Bold"],
                    "text-size": 14,
                    'text-anchor': 'center',
                    'text-offset': [0, 0]
                },
                paint: {
                    'text-color': '#000',
                    'text-halo-color': '#fff',
                    'text-halo-width': 1,
                },
            });
            map.on('sourcedata', () => {
                const features = map.querySourceFeatures('my-data-source', { sourceLayer: 'terrain_wojewodztwa_centroid' });
                features.forEach(feature => {
                    const featureId = feature.id; // Ensure each feature has a unique ID
                    if (featureId) {
                        feature.properties['external'] = 'test' // here would be a value /GET from server
                    }
                });
            });
```
(this doesn't work)
It could load just when source is being loaded, but I can't really put this data in vector source (using Martin with postgis, however my functions would be too complex to create - too many settings to configure with only query params and PG functions)
Also I don't want it to be just points in GeoJSON, as in some cases I'd like to style polygon fill based on same value
Thanks

**Comment by HarelM** [comment]
You can use `setLayoutProperty`, add specific layers for specific features etc.

**Reply by zmatez** [reply]
Well, if I have 2000 features and for each unique value, I don't think that's good idea. Any other possiblities?

**Reply by HarelM** [reply]
I don't know, you'll need to figure this out on your own I guess...

**Comment by wipfli** [comment]
I did a bit of prototyping to support this use-case but have not really finished it yet. You can have a look at the idea here: https://github.com/wipfli/maplibre-feature-properties-transform-example maybe the live weather example is roughly what you are looking for...
